Replaced VMAlert proxy with Alerting tab in VMUI

AndrewChubatiuk opened the pull request:
### Describe Your Changes

Rules page header + content
<img width="1235" height="520" alt="image" src="https://github.com/user-attachments/assets/bb0c5818-c44a-46e6-bc47-e6718be34016" />
Expanded rule without alert
<img width="1418" alt="image" src="https://github.com/user-attachments/assets/ae0b265f-24fe-4549-8913-b1be8e7c2862" />
Expanded rule with alert
<img width="1418" alt="image" src="https://github.com/user-attachments/assets/8a138403-0712-4de2-bfa5-467da3a979dd" />
Notifiers page
<img width="1419" alt="image" src="https://github.com/user-attachments/assets/557c2831-e960-44ec-9b93-f1ebfeb1fbb0" />

fixes https://github.com/VictoriaMetrics/VictoriaMetrics/issues/8330
fixes https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6091
fixes https://github.com/VictoriaMetrics/VictoriaLogs/issues/90

VMUI:
- Added added `Alerting -> Rules` and `Alerting -> Notifiers` pages for VictoriaMetrics 
- Support includeAll option in Select component

VMAlert:
- added `/api/v1/group`useful to get information about certain group
- added `lastError` for `/api/v1/notifiers` for each target to see information about failed notifiers

### Checklist

The following checks are **mandatory**:

- [ ] My change adheres to [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/victoriametrics/contributing/).

arturminchukov commented on app/vmui/packages/vmui/src/components/ExploreAlerts/Notifier/index.tsx:
```diff
@@ -0,0 +1,42 @@
+
+  return (
+    <div className="vm-explore-alerts-notifier">
+      <table>
+        <thead>
+          <tr>
+            <th className="col-small">Address</th>
+            <th>Labels</th>
+          </tr>
+        </thead>
+        <tbody>
+          {targets.map(target => (
+            <tr key={target.address}>
+              <td className="col-small">{target.address}</td>
+              <td className="badge-container">
+                {Object.entries(target.labels || {}).map(([name, value]) => (
+                  <span
+                    className="badge"
+                    key={name}
+                  >{`${name}: ${value}`}</span>
+                ))}
+              </td>
+            </tr>
+          ))}
+        </tbody>
+      </table>
```
Can we use instead of this `Table` component from [`src/components/Table/Table.tsx`](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/777f05efcc3ef3440c8821861dbc91357dd9c586/app/vmui/packages/vmui/src/components/Table/Table.tsx)?

AndrewChubatiuk replied:
haven't checked it, will take a look

AndrewChubatiuk replied:
just checked Table component implementation and it's use cases, I'm not sure it's good for this data, also not sure if table is the best choice for this content at all, maybe for rule annotations only. wanted to reuse something similar to group logs header

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/62e20f92-fa03-432c-9b09-9047ac1400fa" />

arturminchukov commented on app/vmui/packages/vmui/src/components/ExploreAlerts/NotifierHeader/index.tsx:
```diff
@@ -0,0 +1,30 @@
+}) => {
+  const { isMobile } = useDeviceDetect();
+
+  if (isMobile) {
+    return (
+      <div className="vm-explore-alerts-notifier-header vm-explore-alerts-notifier-header_mobile">
+        <div className="vm-explore-alerts-notifier-header__name">{notifier.kind}</div>
+      </div>
+    );
+  }
+
+  return (
+    <div className="vm-explore-alerts-notifier-header">
+      <div className="vm-explore-alerts-notifier-header__name">{notifier.kind}</div>
+    </div>
+  );
+};
```
Maybe will be better to join mobile and desktop version into one
```suggestion
  return (
    <div className={classNames({
        "vm-explore-alerts-notifier-header": true, 
        "vm-explore-alerts-notifier-header_mobile": isMobile
        })}
     >
      <div className="vm-explore-alerts-notifier-header__name">{notifier.kind}</div>
    </div>
  );
};
```

arturminchukov commented on app/vmui/packages/vmui/src/components/ExploreAlerts/Rule/index.tsx:
```diff
@@ -0,0 +1,163 @@
+                  <td className="col-small">{name}</td>
+                  <td>{value}</td>
+                </tr>
+              ))}
+            </tbody>
+          </table>
+          {rule.alerts && (
+            <>
+              <span className="title">Alerts</span>
+              <table>
+                <thead>
+                  <tr>
+                    <th className="col-small">Active since</th>
+                    <th className="col-small">Value</th>
+                    <th>Labels</th>
+                  </tr>
+                </thead>
+                <tbody>
+                  {rule.alerts.map(alert => (
+                    <tr
+                      id={`alert-${alert.id}`}
+                      key={alert.id}
+                      onClick={openLink(alert.sourceLink)}
+                    >
+                      <td className="col-small">{dayjs().to(dayjs(alert.activeAt))}</td>
+                      <td className="col-small">
+                        <div className={`badge ${alert.state}`}>{alert.value}</div>
+                      </td>
+                      <td className="badge-container">
+                        {Object.entries(alert.labels || {}).map(([name, value]) => (
+                          <span
+                            key={name}
+                            className="badge"
+                          >{`${name}: ${value}`}</span>
+                        ))}
+                      </td>
+                    </tr>
+                  ))}
+                </tbody>
+              </table>
```
Is it possible to use here `Table` component from [src/components/Table/Table.tsx](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/777f05efcc3ef3440c8821861dbc91357dd9c586/app/vmui/packages/vmui/src/components/Table/Table.tsx)?

AndrewChubatiuk replied:
unfortunately it's not, visualisation is quite different

arturminchukov commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/ExploreNotifiers.tsx:
```diff
@@ -0,0 +1,153 @@
+    return errorNotifiers;
+  }, [errorNotifiers]);
+
+  if (!isLoading && notifiers && hash && !error) {
+    useEffect(() => {
+      const target = document.querySelector(hash);
+      if (target) {
+        let parent = target.closest("details");
+        while(parent) {
+          parent.open = true;
+          parent = parent.parentElement.closest("details");
+        }
+        target.scrollIntoView();
+      }
+    }, [hash]);
+  }
```
It's better to move condition inside of `useEffect`
```suggestion
    useEffect(() => {
      if (!isLoading && notifiers && hash && !error) {
        const target = document.querySelector(hash);
        if (target) {
          let parent = target.closest("details");
          while(parent) {
            parent.open = true;
            parent = parent.parentElement.closest("details");
          }
          target.scrollIntoView();
        }
      }
    }, [hash, isLoading, notifiers, error]);
```
Because it could throw an error , something like `Hooks can only be called at the top level...` . It is also related to other hooks too.

AndrewChubatiuk commented on app/vmui/packages/vmui/vite.config.ts:
```diff
@@ -10,17 +10,30 @@ const getProxy = (): Record<string, ProxyOptions> | undefined => {
+    case "ALERT": {
+      return {
+        "/api": {
+          target: "http://localhost:8880",
```
it was done intentionally, since this PR introduces endpoints or changes to API, that are not yet available on any playground publicly

AndrewChubatiuk commented:
with vmui it's not possible, since it uses hash router, only redirects are possible here

Loori-R commented:
> Redirects might break setups when access to the alert UI goes through a proxy (like vmauth) with a preconfigured path.

@AndrewChubatiuk Would it be possible to keep the old routes as they are, without redirects, and simply add a link to the new interface on the old pages?

AndrewChubatiuk commented:
there're no old pages, this PR also replaces VMAlert ui with VMUI

AndrewChubatiuk commented on app/vmui/packages/vmui/src/types/index.ts:
```diff
@@ -179,3 +185,65 @@ export interface GroupLogsType {
+  lastSeriesFetched: int64
+  id: int64
+  debug: boolean
+}
```
thought I've already pushed that change

AndrewChubatiuk commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/ExploreNotifiers.tsx:
```diff
@@ -0,0 +1,153 @@
+          filteredTargets.push(target);
+        }
+      }
+    });
```
updated

AndrewChubatiuk commented on app/vmui/packages/vmui/src/hooks/useSearchParamsFromObject.ts:
```diff
@@ -7,11 +7,18 @@ const useSearchParamsFromObject = () => {
+
     Object.entries(objectParams).forEach(([key, value]) => {
-      if (searchParams.get(key) !== `${value}`) {
+      if (!searchParams.get(key) !== `${value}`) {
```
it's already removed

AndrewChubatiuk commented on app/vmui/packages/vmui/package.json:
```diff
@@ -14,11 +14,12 @@
     "marked": "^15.0.8",
     "marked-emoji": "^2.0.0",
     "preact": "^10.26.5",
+    "prismjs": "^1.30.0",
```
removed these packages

AndrewChubatiuk commented on app/vmui/packages/vmui/src/router/navigation.ts:
```diff
@@ -72,17 +64,20 @@ export const getDefaultNavigation = ({
-export const getLogsNavigation = (): NavigationItem[] => [
+export const getLogsNavigation = ({
+  showAlertLink,
+}): NavigationItem[] => [
```
this is outdated as well

Loori-R reviewed (approved):
LGTM. Thanks!

Haleygo commented:
<img width="787" height="167" alt="image" src="https://github.com/user-attachments/assets/21a38c1c-6e36-46ce-99f8-f4b84176fe41" />

Having "Alerts" under the "Alerts" page is confusing, maybe we should rename them both.

And why does the "Alerts" page use `/api/v1/rules?type=alert` instead of `/api/v1/alerts`? How does this differ from the "Rules" page with the "Alerting" rule type filter?

<img width="1458" height="611" alt="image" src="https://github.com/user-attachments/assets/0696acec-0cee-4270-beec-ec7d589a0bb3" />

>Separate tabs for recording rules and alerts.

I agree we should have a separate page for rule details (both alerting and recording rules) using the `/api/v1/rule` API, which could also be used to include more info as mentioned in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/9493.

AndrewChubatiuk commented:
@Haleygo 
will answer your questions from image here:
1. too many tooltips may cause a page to hang from time to time, maybe pagination should be implemented first, but this may complicate searching
2. same as the first
3, 4. will fix it
5. rule expansion doesn't trigger API call, it just shows preloaded hidden content. for this functionality there should be a separate API call, also in this case should already opened rules be updated periodically? suppose this is an improvement for next iteration of UI updates
6. instead of source link anchor each table item is now clickable

Haleygo commented:
@AndrewChubatiuk 

>Don't see much sense in it either, just migrated it from vmalert

It makes sense in the old vmalert UI, since the `Alerts` means the page only shows firing alerts. In VMUI, we should probably change the name of vmalert page entrance.

>Old VMAlert UI actually shows the same data

In old vmalert UI, the list is based on firing alerts, it doesn't show alerting rule that haven't been triggered.

>2. same as the first

I think we can replace the rule type icons with words `Alert` and `Record` with a colorful block(forgive my limited language).

>also in this case should already opened rules be updated periodically? 

I think currently, we don't need to update the time at all, unless the page is refreshed.

>4. instead of source link anchor each table item is now clickable

That is also confusing, easily jump to the source link without any visible tip.

AndrewChubatiuk commented:
- removed `Alerts` page 
- renamed `Alerts` dropdown to `Alerting`
- replaced circles on the right with text, see first image in description
- removed original changes from origin vmalert ui
- added tooltips for Alerting rule and Recording rule icon
- duration is visible only for items with non-zero duration values

AndrewChubatiuk commented on app/vmselect/main.go:
```diff
@@ -492,25 +492,6 @@ func handleStaticAndSimpleRequests(w http.ResponseWriter, r *http.Request, path
 		return true
 	}
 
-	if path == "/vmalert" {
```
initially wanted to rely on /api/v1/rules only, but then understood that other endpoints are needed as well and added back /vmalert prefix. reverted this change

hagen1778 replied:
Oh, it is needed only for proxying requests to vmalert's UI. So not needed now.

AndrewChubatiuk commented on app/vmui/packages/vmui/src/components/Main/Icons/index.tsx:
```diff
@@ -1,5 +1,32 @@
+  </svg>
+);
+
+export const RecordingRuleIcon = () => (
```
now it's tape icon, which associates with recording as well, but can switch to something else. IMO floppy disk associates with Save function

Haleygo commented on app/vmselect/main.go:
```diff
@@ -700,7 +693,6 @@ var (
 	expandWithExprsRequests = metrics.NewCounter(`vm_http_requests_total{path="/expand-with-exprs"}`)
 	prettifyQueryRequests   = metrics.NewCounter(`vm_http_requests_total{path="/prettify-query"}`)
 
-	vmalertRequests   = metrics.NewCounter(`vm_http_requests_total{path="/vmalert"}`)
```
This metric is still used by `/vmalert` prefix requests.

AndrewChubatiuk replied:
fixed

AndrewChubatiuk commented:
- added separate modal for group information
- added lastError to notifier api endpoint and improved error visualisation for both rules and notifiers
- updated some icons to differentiate alerting rules, recording rules and alerts
- removed time formatting
- some other improvements

hagen1778 commented on app/vmui/packages/vmui/src/components/ExploreAlerts/BaseAlert/index.tsx:
```diff
@@ -0,0 +1,89 @@
+                startIcon={<SearchIcon />}
+                onClick={openQueryLink}
+              >
+                <span className="vm-button-text">Show query</span>
```
```suggestion
                <span className="vm-button-text">Run query</span>
```

When built locally, this button seems out of place:
<img width="612" height="172" alt="image" src="https://github.com/user-attachments/assets/c7448a92-42c1-478b-a8fd-27827476ee73" />

Can we move it to the right border of the whole block, below the FIRING/PENDING icons?

hagen1778 reviewed (commented):
Looks good! A few minor comments.

Please add changelog and test to the updated vmaert API.

AndrewChubatiuk commented on app/vmui/packages/vmui/src/components/ExploreAlerts/BaseRule/index.tsx:
```diff
@@ -0,0 +1,201 @@
+          {!!item.duration && (
+            <tr>
+              <td className="vm-col-md">For</td>
+              <td>{formatDuration(item.duration)}</td>
```
updated format, when only seconds are present duration will be displayed as `20s`, with milliseconds - `20.03s`, milliseconds only - `0.001s`

AndrewChubatiuk commented on app/vmui/packages/vmui/src/components/ExploreAlerts/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,125 @@
+            startIcon={<ArrowDownIcon />}
+            onClick={openGroupLink}
+          >
+            <span className="vm-button-text">Rule Group</span>
```
replaced it with `Copy link` button and on modal closing element id is added to url hash to scroll page to just closed item

AndrewChubatiuk commented:
> Please add changelog and test to the updated vmaert API.
done

AndrewChubatiuk commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/ExploreNotifiers.tsx:
```diff
@@ -0,0 +1,183 @@
+        window.scrollTo(0, parseInt(savedScrollTop));
+      }
+      const handleBeforeUnload = () => {
+        localStorage.setItem("scrollTop", (window.scrollY || 0).toString());
```
removed storing accordion items state locally

hagen1778 commented on app/vmui/packages/vmui/src/components/ExploreAlerts/ItemHeader/index.tsx:
```diff
@@ -0,0 +1,127 @@
+            startIcon={<LinkIcon />}
+            onClick={copyLink}
+          >
+            <span className="vm-button-text">Copy Link</span>
```
The button now seems copying incorrect URL.
<img width="758" height="435" alt="image" src="https://github.com/user-attachments/assets/58158f24-aa18-4661-afcc-9b65b84cd2ec" />

The copy gets the following link `http://localhost:8428/#/rules?group_id=15701237453163353695&alert_id=7888404066739011861`

This link suppose to open this alert again when clicked, but it does not:
<img width="872" height="460" alt="image" src="https://github.com/user-attachments/assets/e448112b-bd83-42ed-98f5-b3a777172eef" />

AndrewChubatiuk commented on app/vmui/packages/vmui/ExecutionControls:
```diff
@@ -0,0 +1,33 @@
+# This viminfo file was generated by Vim 9.1.
```
have no idea how it appeared here, it's not needed

hagen1778 commented on app/vmselect/main.go:
```diff
@@ -492,13 +492,6 @@ func handleStaticAndSimpleRequests(w http.ResponseWriter, r *http.Request, path
 		return true
 	}
 
-	if path == "/vmalert" {
```
After this change, links like `http://localhost:8428/vmalert` will stop working. If you rever this change, links like this will continue working, as expected.

arturminchukov commented on app/vmui/Dockerfile-build:
```diff
@@ -1,4 +1,4 @@
-FROM node:20-alpine3.19
+FROM node:22-alpine3.22
```
It's ok.

arturminchukov commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/ExploreRules.tsx:
```diff
@@ -0,0 +1,244 @@
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  if (!location.hash && groupId) {
+    setModalOpen(true);
+  } else {
+    setModalOpen(false);
+  }
+
```
move it to `useEffect` 
```suggestion
  useEffect(() => {
    if (!location.hash && groupId) {
      setModalOpen(true);
    } else {
      setModalOpen(false);
    }
  }, [location.hash, groupId]);

```

arturminchukov commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/ExploreRules.tsx:
```diff
@@ -0,0 +1,244 @@
+
+  groups.forEach((group) => {
+    const filteredRules: APIRule[] = [];
+    const statesPerGroup: Record<string, number> = {};
+    group.rules.forEach((rule) => {
+      const ruleType = rule.type.charAt(0).toUpperCase() + rule.type.slice(1);
+      allTypes.add(ruleType);
+      if (types?.length && !types.includes(ruleType)) return;
+
+      const state = getState(rule);
+      const stateName = state.charAt(0).toUpperCase() + state.slice(1);
+      allStates.add(stateName);
+      if (states?.length && !states.includes(stateName)) return;
+
+      if (
+        searchInput &&
+        !rule.name.toLowerCase().includes(searchInput.toLowerCase()) &&
+        !group.name.toLowerCase().includes(searchInput.toLowerCase()) &&
+        !group.file.toLowerCase().includes(searchInput.toLowerCase())
+      )
+        return;
+
+      filteredRules.push(rule);
+      if (state !== "no match" && state !== "unhealthy" && state !== "firing" && state !== "pending")
+        return;
+
+      const count = state === "firing" || state === "pending" ? rule?.alerts?.length : 1;
+      if (stateName in statesPerGroup) {
+        statesPerGroup[stateName] += count;
+      } else {
+        statesPerGroup[stateName] = count;
+      }
+    });
+    if (filteredRules.length) {
+      const g = Object.assign({}, group);
+      g.rules = filteredRules;
+      g.states = statesPerGroup;
+      filteredGroups.push(g);
+    }
+  });
```
Maybe we can move this logic to some `utils` file, and here use function call, as example:
```suggestion
  const { filteredGroups, allTypes, allStates } = useMemo(
     () => filterGroups(groups || [], types, states, searchInput),
  [groups, types, states, searchInput]
);
```
As for me it will be better for readability.

arturminchukov commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/hooks/useSetQueryParams.ts:
```diff
@@ -0,0 +1,70 @@
+    setSearchParamsFromKeys(params);
+  };
+
+  useEffect(setSearchParamsFromState, [kinds, search]);
+  useEffect(setSearchParamsFromState, []);
```
unnecessary `useEffect`, the same behavior will be on `useEffect(setSearchParamsFromState, [kinds, search]);`, so it can be removed 
```suggestion
  useEffect(setSearchParamsFromState, [kinds, search]);
```

arturminchukov commented on app/vmui/packages/vmui/src/pages/ExploreAlerts/hooks/useSetQueryParams.ts:
```diff
@@ -0,0 +1,70 @@
+    setSearchParamsFromKeys(params);
+  };
+
+  useEffect(setSearchParamsFromState, [
+    types,
+    states,
+    search,
+    rule_id,
+    group_id,
+    alert_id,
+  ]);
+  useEffect(setSearchParamsFromState, []);
```
unnecessary `useEffect(setSearchParamsFromState, []);`
```suggestion
  useEffect(setSearchParamsFromState, [
    types,
    states,
    search,
    rule_id,
    group_id,
    alert_id,
  ]);
```

arturminchukov reviewed (changes requested):
Looks good! just several comments

hagen1778 reviewed (approved):
Thanks!

hagen1778 commented:
@AndrewChubatiuk see follow-up in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/9674

valyala commented:
@hagen1778 , @makasim , please see https://github.com/VictoriaMetrics/VictoriaLogs/pull/522#issuecomment-3304534286
